lib/filewatch/read_mode/handlers/read_file.rb

```diff
@@ -5,42 +5,51 @@ class ReadFile < Base
-            changed = true
-            result.lines.each do |line|
-              watched_file.listener.accept(line)
-              # sincedb position is independent from the watched_file bytes_read
-              sincedb_collection.increment(watched_file.sincedb_key, line.bytesize + @settings.delimiter_byte_size)
-            end
-            sincedb_collection.request_disk_flush
-          rescue EOFError
-            # flush the buffer now in case there is no final delimiter
-            line = watched_file.buffer.flush
-            watched_file.listener.accept(line) unless line.empty?
-            watched_file.listener.eof
-            watched_file.file_close
-            key = watched_file.sincedb_key
-            sincedb_collection.reading_completed(key)
-            sincedb_collection.clear_watched_file(key)
-            watched_file.listener.deleted
-            watched_file.unwatch
-            break
-          rescue Errno::EWOULDBLOCK, Errno::EINTR
-            watched_file.listener.error
-            break
-          rescue => e
-            logger.error("read_to_eof: general error reading file", "path" => watched_file.path, "error" => e.inspect, "backtrace" => e.backtrace.take(8))
-            watched_file.listener.error
-            break
+          loop_control = watched_file.loop_control_adjusted_for_stat_size
+          controlled_read(watched_file, loop_control)
+          sincedb_collection.request_disk_flush
+          break unless loop_control.more
+        end
+        if watched_file.all_read?
+          # flush the buffer now in case there is no final delimiter
+          line = watched_file.buffer.flush
+          watched_file.listener.accept(line) unless line.empty?
+          watched_file.listener.eof
+          watched_file.file_close
+          key = watched_file.sincedb_key
+          sincedb_collection.reading_completed(key)
+          sincedb_collection.clear_watched_file(key)
+          watched_file.listener.deleted
+          watched_file.unwatch
+        end
+      end
+    end
+
+    def controlled_read(watched_file, loop_control)
+      logger.trace("reading...", "iterations" => loop_control.count, "amount" => loop_control.size, "filename" => watched_file.filename)
+      loop_control.count.times do
+        begin
+          result = watched_file.read_extract_lines(loop_control.size) # expect BufferExtractResult
+          logger.info(result.warning, result.additional) unless result.warning.empty?
+          result.lines.each do |line|
+            watched_file.listener.accept(line)
+            # sincedb position is independent from the watched_file bytes_read
+            delta = line.bytesize + @settings.delimiter_byte_size
+            sincedb_collection.increment(watched_file.sincedb_key, delta)
           end
+        rescue EOFError
+          logger.error("controlled_read: eof error reading file", "path" => watched_file.path, "error" => e.inspect, "backtrace" => e.backtrace.take(8))
+          break
```

Comment:
noticing there is no `watched_file.listener.error` for this exception, I am assuming this is on purpose?

Reply:
It is. As this is a controlled read to the end of the file as given by the `stat` size, EOF should never happen. Currently the listener does nothing with `eof` and `error`. I debated with myself whether to evict the multiline codec from the identity map on eof before but decided against it at the time - perhaps now, because of controlled read and `read mode`, signalling eof to the listener should evict the identity.